### ISSUE TITLE
MarkovChain.simulate can sample invalid states

### ISSUE THREAD

**Issue by btanner** [QuantEcon/QuantEcon.py#591]
Apologies, I'm quite new to this project, so there may well be some overarching design decisions on input validation or error reporting that make this irrelevant.

### TL;DR
 It is possible to create a MarkovChain with a P matrix that passes validation, but does not exactly have all transition probabilities rows sum to 1, which can cause the simulate method to silently transition to an invalid absorbing state.

### Detailed Description
The validation is [here in the code](https://github.com/QuantEcon/QuantEcon.py/blob/267a9bb723b5db163c02e57971a50284ab4e3545/quantecon/markov/core.py#L196):
```
row_sums = self.P.sum(axis=1)
...
if not np.allclose(row_sums, np.ones(self.n)):
   raise ValueError('The rows of P must sum to 1')
```

We found this when we were creating transition matrices using JAX, which defaults to float32 precision so some rows just can't quite sum to exactly 1. For the moment, I'll focus on dense matrices.

If the rows do not quite sum to 1, then the cdfs won't look quite right and will have a small bit of unaccounted probability mass from [this line](https://github.com/QuantEcon/QuantEcon.py/blob/267a9bb723b5db163c02e57971a50284ab4e3545/quantecon/markov/core.py#L424):
```
np.cumsum(self.P, axis=-1, out=cdfs)
```

If you try to simulate a trajectory, you will call [simulate](https://github.com/QuantEcon/QuantEcon.py/blob/267a9bb723b5db163c02e57971a50284ab4e3545/quantecon/markov/core.py#L529)  --> [simulate indices](https://github.com/QuantEcon/QuantEcon.py/blob/267a9bb723b5db163c02e57971a50284ab4e3545/quantecon/markov/core.py#L441) --> [_generate_sample_paths](https://github.com/QuantEcon/QuantEcon.py/blob/267a9bb723b5db163c02e57971a50284ab4e3545/quantecon/markov/core.py#L578) --> [searchsorted](https://github.com/QuantEcon/QuantEcon.py/blob/267a9bb723b5db163c02e57971a50284ab4e3545/quantecon/util/array.py#L18).

The problem happens in `searchsorted` and the behaviour of that function is as advertised.

Notably, `simulate_paths` does some checking to ensure the starting states are valid, but it does not check to make sure the generated paths are valid.

### Fixes
I see a few potential fixes.

1. Tighten validation in MarkovChain constructor. Instead of np.allclose, potentially reject if rows sums aren't exactly 1.0
2. Upscale precision and re-normalize in MarkovChain constructor to ensure they sum to 1.0. I'm not _sure_ if this will work 100% of the time.
3. Notice when searchsorted produces invalid states and resample.
4. Log an error or raise an exception when simulate generates invalid states.

These all have pros and cons.  We could prepare a PR for one of these if it was supported by your project team. My preferred strategy would be either (2) or (3).

(2) Might just be the simplest thing to do. Perhaps log a message if the step is necessary.

(3) Could be done a few different ways. The simplest would be to just reject everything if any state is invalid. But, you could be more careful for sure.

Any thoughts?

**Comment by oyamad** [comment]
@btanner Thank you for opening this issue.

To better understand the problem, can you share your code creating your transition matrices?

**Comment by btanner** [comment]
@oyamad  - I'm a little unsure why this repro is non-deterministic, I think I have controlled for all the randomness. But anyway, here is a "search for the case" style repro and some sample output.

Basically, it creates a simple Markov Chain where State 0 goes to a few other states, and all other states deterministically return to State 0. There is a small bit of missing mass in the CDF, and once in a while you'll catch it.

```
import quantecon
import numpy as np

num_states = 5
transition_states = 4

transition_seed = 1
while True:
  np.random.seed(transition_seed)

  transitions = np.random.uniform(0., 1., transition_states)
  transitions /= np.sum(transitions)
  if np.sum(transitions) != 1.:
    print(f'Transitions {transitions} sum to: {np.sum(transitions)} w/ seed:{transition_seed}')
    P = np.zeros((num_states, num_states), dtype=np.float32)
    P[0,:transition_states] = transitions
    P[1:, 0] = 1.
    print(f'Transition matrix:\n{P}')

    mc = quantecon.markov.MarkovChain(P=P)
    print(f'CDFs:\n{mc.cdfs}')

    simulate_seed = 1
    while True:
      ts_length = 10000
      seq = mc.simulate(ts_length=ts_length, init=0, num_reps=1, random_state=simulate_seed)
      max_state_in_seq = np.max(seq)
      if max_state_in_seq > transition_states - 1:
        print(f'In sequence from seed:{simulate_seed}, ended up in invalid state:{max_state_in_seq}')
        time_in_invalid_state = np.count_nonzero(np.where(seq==max_state_in_seq, 1, 0))
        print(f'In that state for {time_in_invalid_state} of {ts_length} steps')
        break
      simulate_seed += 1
    break
  transition_seed += 1
```

Sample Output:
```
Transitions [0.30132911 0.01791839 0.37988817 0.30086432] sum to: 1.0000000000000002 w/ seed:2
Transition matrix:
[[0.3013291  0.01791839 0.37988818 0.3008643  0.        ]
 [1.         0.         0.         0.         0.        ]
 [1.         0.         0.         0.         0.        ]
 [1.         0.         0.         0.         0.        ]
 [1.         0.         0.         0.         0.        ]]
CDFs:
[[0.30132911 0.3192475  0.69913568 0.99999999 0.99999999]
 [1.         1.         1.         1.         1.        ]
 [1.         1.         1.         1.         1.        ]
 [1.         1.         1.         1.         1.        ]
 [1.         1.         1.         1.         1.        ]]
In sequence from seed:22220, ended up in invalid state:5
In that state for 4645 of 10000 steps
```

**Comment by oyamad** [comment]
@btanner PR #592 fixes your *specific* instance.

I am not much knowledgable about complications in floats, but it seems to be not  a good idea to change the `dtype` in creating the `cdfs` array (the previous code creates it with the default `dtype` which is in most cases `np.float64` irrespective of the `dtype` of the original `P` which is `np.float32` in your case).

Please let us know if you find an instance that does not work with this fix.

**Comment by mmcky** [comment]
@btanner we have issued a new release `0.5.2` which includes the bug fix above.

**Comment by btanner** [comment]
Thanks for your attention here @oyamad and @mmcky. Definitely did not expect such a quick response and turnaround!

I hacked this change into my local example and it does resolve the issue for that particular seed, but I can can find other examples that fail.

This is not something that you have to address in your project, but I do think some extra context is helpful.

In machine learning, folks are often using frameworks like PyTorch, TensorFlow, or JAX to do accelerated calculations. All of these frameworks use float32 by default, and can even use bfloat16. The precision of these data types if much less than float64, and they simply cannot represent the same numbers.

float64 is of course not magic either, there are limits to what it can represent as well. We can see that here:
```
>>> print(np.sum(np.ones(3)/3))
1.0
>>> print(np.sum(np.ones(7)/7))
0.9999999999999998
```

So, over time, it's possible that people will generate values using these tools and then pass them off to your library. My only concern here is that since I got "caught" by this, it would be nice if it didn't happen to others.

I think, if you don't want to re-normalize P, it might be nice to just check the dtype and throw a ValueError when creating a MarkovChain with dtype != float64 to ensure that folks are not caught by this precision problem.

On our side, we have already changed our code to upscale precision to float64 and renormalize. We're also checking for invalid states returned from MarkovChain.simulate, and we'll resample if it happens.

I'm also running a search to see if I can find a failure case of Simulate using float64. So far, so good (no problems), but I'll let you know if I find anything.

**Comment by oyamad** [comment]
Let me close this. Please reopen if any problem has been found.
